### ISSUE TITLE
resolved is randomly crashing (hashmap_iterate_entry())

### ISSUE THREAD

**Issue by bigon** [systemd/systemd#16168]
Hello,

On debian unstable, with systemd 245.6-1, resolved crashed multiple times with the following trace:

```
           PID: 338706 (systemd-resolve)
           UID: 102 (systemd-resolve)
           GID: 104 (systemd-resolve)
        Signal: 11 (SEGV)
     Timestamp: Sat 2020-06-13 08:07:53 CEST (20min ago)
  Command Line: /lib/systemd/systemd-resolved
    Executable: /usr/lib/systemd/systemd-resolved
 Control Group: /system.slice/systemd-resolved.service
          Unit: systemd-resolved.service
         Slice: system.slice
       Boot ID: 73d56b833b174db5814a013cf44aaf57
    Machine ID: eb71526a9be548c2806ab24bb024fb1f
      Hostname: edoras
       Storage: /var/lib/systemd/coredump/core.systemd-resolve.102.73d56b833b174db5814a013cf44aaf57.338706.1592028473000000000000.lz4
       Message: Process 338706 (systemd-resolve) of user 102 dumped core.
                
                Stack trace of thread 338706:
                #0  0x00007fdd53305b2a hashmap_iterate_entry (libsystemd-shared-245.so + 0x110b2a)
                #1  0x00007fdd53307a52 internal_hashmap_iterate (libsystemd-shared-245.so + 0x112a52)
                #2  0x000055c391b33b17 dns_query_candidate_go (systemd-resolved + 0x32b17)
                #3  0x000055c391b3312c dns_query_candidate_notify (systemd-resolved + 0x3212c)
                #4  0x000055c391b33388 dns_transaction_complete (systemd-resolved + 0x32388)
                #5  0x000055c391b2bccd dns_transaction_process_dnssec (systemd-resolved + 0x2accd)
                #6  0x000055c391b2cb2d dns_transaction_process_reply (systemd-resolved + 0x2bb2d)
                #7  0x000055c391b2d41b on_dns_packet (systemd-resolved + 0x2c41b)
                #8  0x00007fdd53279ac7 source_dispatch (libsystemd-shared-245.so + 0x84ac7)
                #9  0x00007fdd53279d61 sd_event_dispatch (libsystemd-shared-245.so + 0x84d61)
                #10 0x00007fdd5327b798 sd_event_run (libsystemd-shared-245.so + 0x86798)
                #11 0x00007fdd5327b9b7 sd_event_loop (libsystemd-shared-245.so + 0x869b7)
                #12 0x000055c391b0cb7a run (systemd-resolved + 0xbb7a)
                #13 0x00007fdd5347ee0b __libc_start_main (libc.so.6 + 0x26e0b)
                #14 0x000055c391b0d4ca _start (systemd-resolved + 0xc4ca)
```

In dmesg, I can see: `[605459.479558] traps: systemd-resolve[338706] general protection fault ip:7fdd53305b2a sp:7fff14895a10 error:0 in libsystemd-shared-245.so[7fdd53245000+170000]`

**Comment by poettering** [comment]
hmm, what is 245.6? does that contain any backported resolved stuff from git master? @mbiebl?

**Comment by mbiebl** [comment]
https://github.com/systemd/systemd-stable/commits/v245.6

**Comment by poettering** [comment]
is this reproducible? do you have dnssec mode on?

**Comment by poettering** [comment]
how does your search domain list look like when this happens?

**Comment by bigon** [comment]
Yes it's still happening from times to time (2-3 times a month) but the trace is a bit different:

```
           PID: 640 (systemd-resolve)
           UID: 102 (systemd-resolve)
           GID: 104 (systemd-resolve)
        Signal: 11 (SEGV)
     Timestamp: Mon 2020-11-09 12:42:09 CET (1 weeks 2 days ago)
  Command Line: /lib/systemd/systemd-resolved
    Executable: /usr/lib/systemd/systemd-resolved
 Control Group: /system.slice/systemd-resolved.service
          Unit: systemd-resolved.service
         Slice: system.slice
       Boot ID: 2179fbffca754f02a25b768c7d5c23f3
    Machine ID: eb71526a9be548c2806ab24bb024fb1f
      Hostname: edoras
       Storage: /var/lib/systemd/coredump/core.systemd-resolve.102.2179fbffca754f02a25b768c7d5c23f3.640.1604922129000000.zst (inaccessible)
       Message: Process 640 (systemd-resolve) of user 102 dumped core.
                
                Stack trace of thread 640:
                #0  0x00007fd0fb5a4dd3 _hashmap_iterate (libsystemd-shared-246.so + 0x114dd3)
                #1  0x00005628ce31eea7 dns_query_candidate_go (systemd-resolved + 0x23ea7)
                #2  0x00005628ce31edbc dns_query_candidate_notify (systemd-resolved + 0x23dbc)
                #3  0x00005628ce31ff48 dns_transaction_complete (systemd-resolved + 0x24f48)
                #4  0x00005628ce327f81 dns_transaction_process_reply (systemd-resolved + 0x2cf81)
                #5  0x00005628ce328ca3 on_dns_packet (systemd-resolved + 0x2dca3)
                #6  0x00007fd0fb639dc7 source_dispatch (libsystemd-shared-246.so + 0x1a9dc7)
                #7  0x00007fd0fb63b0c1 sd_event_dispatch (libsystemd-shared-246.so + 0x1ab0c1)
                #8  0x00007fd0fb63b418 sd_event_run (libsystemd-shared-246.so + 0x1ab418)
                #9  0x00007fd0fb63b637 sd_event_loop (libsystemd-shared-246.so + 0x1ab637)
                #10 0x00005628ce31240d run (systemd-resolved + 0x1740d)
                #11 0x00005628ce306836 main (systemd-resolved + 0xb836)
                #12 0x00007fd0fae5dcca __libc_start_main (libc.so.6 + 0x26cca)
                #13 0x00005628ce3068da _start (systemd-resolved + 0xb8da)
```
resolv.conf looks like:

```
# This file is managed by man:systemd-resolved(8). Do not edit.
#
# This is a dynamic resolv.conf file for connecting local clients to the
# internal DNS stub resolver of systemd-resolved. This file lists all
# configured search domains.
#
# Run "resolvectl status" to see details about the uplink DNS servers
# currently in use.
#
# Third party programs should typically not access this file directly, but only
# through the symlink at /etc/resolv.conf. To manage man:resolv.conf(5) in a
# different way, replace this symlink by a static file or a different symlink.
#
# See man:systemd-resolved.service(8) for details about the supported modes of
# operation for /etc/resolv.conf.

nameserver 127.0.0.53
options edns0 trust-ad
search lan ops.<compagny>.be <compagny>.be
```
Note that the two last entries of the search path are only added when I'm connected to the VPN, I cannot really tell if I was connected at the time of the crash.

I also have the `resolve` NSS module loaded in `nsswitch.conf`:

`hosts:          files mdns4_minimal [NOTFOUND=return] resolve [!UNAVAIL=return] dns myhostname mymachines`

DNSSEC is off on debian builds now

Version I've now is 246.6-3

**Comment by keszybz** [comment]
Fedora bug: https://bugzilla.redhat.com/show_bug.cgi?id=1895937.

**Comment by bigon** [comment]
> Fedora bug: https://bugzilla.redhat.com/show_bug.cgi?id=1895937.

The bug is private BTW

**Comment by keszybz** [comment]
Yeah, that is on purpose. Coredumps contain some private information. If you have a bugzilla account, please send me your address and I can add you to the bug.

**Comment by keszybz** [comment]
Please enable debug logging and when if crashes, attach the logs from before the crash.
(Call `sudo resolvectl log-level debug` for temporary change, or add `Environment=SYSTEMD_LOG_LEVEL=debug` in the [Service] section in the unit file to make the change permanent.)

**Comment by abatkin** [comment]
I have coredumps and debug logs from a fresh and updated install of Fedora 33 (systemd-246.7-2.fc33.x86_64). What's the best way to make those available? @bigon Feel free to add me to that bug (adam at batkin dot net)

I'm seeing systemd-resolve crashes a few times an hour during interactive use. Nothing else on the system is having an issue.

Here's a quick extract from the system journal:
```
Dec 22 12:13:13 xxxx audit[34869]: ANOM_ABEND auid=4294967295 uid=193 gid=193 ses=4294967295 subj=system_u:system_r:systemd_resolved_t:s0 pid=34869 comm="systemd-resolve" exe="/usr/lib/systemd/systemd-resolved" sig=1>
Dec 22 12:13:13 xxxx kernel: traps: systemd-resolve[34869] general protection fault ip:7f2929b85f47 sp:7ffd246b4eb0 error:0 in libsystemd-shared-246.so[7f2929ab2000+1a8000]
Dec 22 12:13:13 xxxx audit: BPF prog-id=95 op=LOAD
Dec 22 12:13:13 xxxx audit: BPF prog-id=96 op=LOAD
Dec 22 12:13:13 xxxx audit: BPF prog-id=97 op=LOAD
Dec 22 12:13:13 xxxx systemd[1]: Started Process Core Dump (PID 35436/UID 0).
Dec 22 12:13:13 xxxx audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=systemd-coredump@11-35436-0 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? >
Dec 22 12:13:13 xxxx systemd-coredump[35443]: Process 34869 (systemd-resolve) of user 193 dumped core.

Stack trace of thread 34869:
    #0  0x00007f2929b85f47 _hashmap_iterate (libsystemd-shared-246.so + 0x129f47)
    #1  0x000056372fab780f dns_query_candidate_go (systemd-resolved + 0x2780f)
    #2  0x000056372fab7720 dns_query_candidate_notify (systemd-resolved + 0x27720)
    #3  0x000056372fabbe28 dns_transaction_complete (systemd-resolved + 0x2be28)
    #4  0x000056372fabdefb dns_transaction_process_reply (systemd-resolved + 0x2defb)
    #5  0x000056372fabed41 on_dns_packet (systemd-resolved + 0x2ed41)
    #6  0x00007f2929c2e217 source_dispatch (libsystemd-shared-246.so + 0x1d2217)
    #7  0x00007f2929c2e4fd sd_event_dispatch (libsystemd-shared-246.so + 0x1d24fd)
    #8  0x00007f2929c2f7c8 sd_event_run (libsystemd-shared-246.so + 0x1d37c8)
    #9  0x00007f2929c2f9ef sd_event_loop (libsystemd-shared-246.so + 0x1d39ef)
    #10 0x000056372faa9c14 run (systemd-resolved + 0x19c14)
    #11 0x000056372fa9e60a main (systemd-resolved + 0xe60a)
    #12 0x00007f29294041e2 __libc_start_main (libc.so.6 + 0x281e2)
    #13 0x000056372fa9e6ae _start (systemd-resolved + 0xe6ae)
Dec 22 12:13:13 xxxx systemd[1]: systemd-coredump@11-35436-0.service: Succeeded.
```

**Comment by keszybz** [comment]
@abatkin please attach them in bugzilla (you should have access now), or send them to me (zbyszek at in waw pl).

**Comment by abatkin** [comment]
Fun story: The core dump/logs that I posted to that Bugzilla are from a "new" PC with a fresh install (AMD 5600X CPU on B550 chipset, so motherboard vendors and Kernel maintainers are still working out some kinks, which made me worried). I just finally upgraded my "old" PC (Intel-based, built in 2014) from Fedora 32 to 33, and now it sees the same (relatively frequent) systemd-resolve crashes.

The only similarity between the who PCs is the network/DNS settings. So maybe something to do with my DNS "server"? My local DNS points at my router (Mikrotik) which points upstream at 8.8.8.8/8.8.4.4.

I'm happy to do additional troubleshooting or even a screen share with someone if that helps to narrow down what might be going on.

**Comment by abatkin** [comment]
I can now reliably reproduce this on my "old" PC (which has been running reliably for 6+ years, i.e. I'm not concerned about memory errors like I am with the new PC). Firefox's startpage is set to `http://startpage.lan/` which is a "local" DNS name served up by my router. I can `dig` just fine (even if I query for ANY, A, AAAA, CNAME directly) but whenever Firefox starts, it causes `systemd-resolve` to crash. The name does resolve, though the host itself (startpage.lan) isn't up.

**Comment by keszybz** [comment]
@abatkin, @bigon it'd be great if you could test the patch in #18290. I like my analysis in the commit message, but I haven't been able to reproduce the issue, so I haven't been able to test it either ;).

**Comment by abatkin** [comment]
That did not do what I had hoped. Once I applied that patch and rebuilt, "nothing worked" (though systemd-resolve stopped crashing). I need to clarify what "nothing worked" means: From a terminal, I could do dns lookups (using `dig`) and I could even `ping` hosts directly. But from Firefox (which has DNS over HTTPS *disabled*) pages would not load - browser windows would just hang. Same with `curl` from the command line (even with `curl -v` it just kinda sat there doing nothing).

A few notes:
* I'm on Fedora 33. My "normal" systemd is `systemd-246.7-2.fc33.x86_64`
* I rebuilt by fetching the SRPM for that specific version and applying your patch, *nothing else*
* I installed the "new" RPMs using `dnf reinstall` and replaced all of the local RPMs that I had installed with the newly built ones (note that I didn't build for i686 so I did not/could not replace `systemd-libs-246.7-2.fc33.i686`)
* To recover my system (since I couldn't make `dnf` download anything either) I downloaded each of the "pristine" RPMs on another system and copied them over with a USB stick (that was silly: I could have just rebuilt - see below - or `scp`'d them)
* In order to make sure my "build process" wasn't the culprit, I commented out the single patch file and followed these steps (which in theory should result in me building RPMs that were roughly identical to the stock Fedora RPMs) and installed those, and everything was back to "normal" (DNS/network all worked properly, systemd-resolve crashed periodically)

If your patch depends on some behavior/changes in the systemd codebase _since_ v246, that could explain what's going on. Otherwise I fear this patch made something worse. Or exposed some behavior that was only happening intermittently (consistent with use-after-free from your description since that may work sometimes and not others) but since you were explicitly checking return values now, whatever "broken" behavior now happens every time (except instead of crashing with your patch, it does something else...)

Anyway, I hope that made some sense - it did in my head.

**Comment by abatkin** [comment]
I should also clarify: After installing a new systemd, I would always reboot before performing any tests.

**Comment by vcaputo** [comment]
Anything in the logs when malfunctioning with the patch applied?

**Comment by abatkin** [comment]
[curl-google-hung-2.txt](https://github.com/systemd/systemd/files/5827578/curl-google-hung-2.txt)

See attached. It's...weird. `dig` works 100% of the time (no explicit nameserver, so it hits `127.0.0.53`). `curl` works almost never. Except when it randomly works. For example, I couldn't get it to talk to https://www.google.com/, I walked away for an hour, came back, ran it again, then it worked. Ran it 3 seconds later and it hung again.

I've got lots of logs, but that's because my system is full of "lots of stuff" and it's not easy to turn everything off when you are in a graphical session (is single user mode with networking even a thing anymore, with systemd? I feel like my unix knowledge from 20 years ago has all become outdated). So attached is a specific sequence of evens that I thought might be helpful (it's annotated in the log). I'm happy to run more tests and upload whatever, but it's just not clear where I should start and end (or what I should do in-between). In the past hour (where I've only been sitting at the keyboard for like 10 minutes) it's generated ~7000 lines.
